### PR TITLE
8308715: Create a mechanism for Implicitly Declared Class javadoc

### PR DESCRIPTION
**Pull request by pavelrappo**
Please review this PR to support _JEP 463 Implicitly Declared Classes and Instance Main Method (Second Preview)_ in JavaDoc.

[JEP 463](https://openjdk.org/jeps/463) is the next iteration of [JEP 445](https://openjdk.org/jeps/445), which introduced the ability to have a source file as a launchable, "classless" method bag

```
% cat HelloWorld.java
/** Run me. */
void main() {
    print("Hello, world!");
}

/** Shortcut for printing strings. */
void print(String arg) {
    System.out.println(arg);
}
```

which the compiler interprets as a familiar class

```
final class HelloWorld {

    HelloWorld() {
    }

    /** Run me. */
    void main() {
        print("Hello, world!");
    }

    /** Shortcut for printing strings. */
    void print(String arg) {
        System.out.println(arg);
    }    
}
```

### How JEP 445 works with JavaDoc today

In JDK 21, javadoc can document such a file **without any changes to the javadoc tool**. The only thing that the user needs to do is to make sure that the following options are present:

* `--enable-preview` and `--source=21`
* `-package`

The first pair of options tells javadoc to use preview features, which JEP 445 is one of. Without these preview-related options, javadoc will raise the following error:

```
% javadoc --version
javadoc 21

% javadoc HelloWorld.java -d /tmp/throwaway
Loading source file HelloWorld.java...
HelloWorld.java:2: error: unnamed classes are a preview feature and are disabled by default.
void main() {
^
  (use --enable-preview to enable unnamed classes)
1 error
```

The second option, `-package`, tells javadoc to document classes that are public, protected, or declared with package access (colloquially known as "package private"). Without this option, javadoc will only document public and protected classes, which do not include the interpreted class:

```
% javadoc --enable-preview --source=21 HelloWorld.java -d /tmp/throwaway
Loading source file HelloWorld.java...
Constructing Javadoc information...
error: No public or protected classes found to document.
1 error
```

When those additional options are present, javadoc does its job:

```
% javadoc --enable-preview --source=21 -package HelloWorld.java -d /tmp/throwaway
Loading source file HelloWorld.java...
Constructing Javadoc information...
Creating destination directory: "/tmp/throwaway/"
Building index for all the packages and classes...
Standard Doclet version 21+35-LTS-2513
Building tree for all the packages and classes...
Generating /tmp/throwaway/HelloWorld.html...
HelloWorld.java:7: warning: no @param for arg
void print(String arg) {
     ^
HelloWorld.java:2: warning: no comment
void main() {
     ^
HelloWorld.java:2: warning: use of default constructor, which does not provide a comment
void main() {
     ^
Generating /tmp/throwaway/package-summary.html...
Generating /tmp/throwaway/package-tree.html...
Generating /tmp/throwaway/overview-tree.html...
Building index for all classes...
Generating /tmp/throwaway/allclasses-index.html...
Generating /tmp/throwaway/allpackages-index.html...
Generating /tmp/throwaway/index-all.html...
Generating /tmp/throwaway/search.html...
Generating /tmp/throwaway/index.html...
Generating /tmp/throwaway/help-doc.html...
3 warnings
```

However, the result does not feel quite right. Firstly, `-package` is too coarse. It includes all top-level classes and their elements, not just the implicit class from `HelloWorld.java`, its default constructor and methods, which are all declared with package access. Secondly, `HelloWorld.java` isn't a first-class citizen in javadoc. That latter fact can be seen from examining stdout and the output directory:

1. DocLint (compiler and javadoc) as well as javadoc itself issue unjust warnings: neither the implicit class nor its default constructor can be documented. The author either does not know about classes and constructors yet (on-ramp audience) or does not care about them (scripts/utilities audience).

   Additionally, because the class' AST node is at the same position as that of the first method declaration, the warning about the undocumented class can be confused with a warning on the first method being undocumented.

2. While such a file is documented as if it were an explicitly declared (normal) class, we might want to dispense with the documentation for the default constructor as it lacks a comment and is an artefact.

### What this PR proposes for JEP 463

1. Leave `--enable-preview` and `--source` as correct and unavoidable until the feature is standardised.
2. "Drill a hole" in javadoc access control to **automatically** allow implicit classes and their public, protected or declared with package access members in documentation.
3. Do not emit warnings for an implicit class and its deault constructor.
4. Do not document an implicit class' default constructor.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change requires CSR request [JDK-8321030](https://bugs.openjdk.org/browse/JDK-8321030) to be approved
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8308715](https://bugs.openjdk.org/browse/JDK-8308715): Create a mechanism for Implicitly Declared Class javadoc (**Enhancement** - P4)
 * [JDK-8321030](https://bugs.openjdk.org/browse/JDK-8321030): Create a mechanism for Implicitly Declared Class javadoc (**CSR**)

### Reviewers
 * [Jonathan Gibbons](https://openjdk.org/census#jjg) (@jonathan-gibbons - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16853/head:pull/16853` \
`$ git checkout pull/16853`

Update a local copy of the PR: \
`$ git checkout pull/16853` \
`$ git pull https://git.openjdk.org/jdk.git pull/16853/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16853`

View PR using the GUI difftool: \
`$ git pr show -t 16853`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16853.diff">https://git.openjdk.org/jdk/pull/16853.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16853#issuecomment-1829982055)

### CODE REVIEW

**Comment by pavelrappo** [comment]
Reviewers, please ignore changes to the following files as well as commits that brought them:

 * .github/workflows/main.yml
 * src/hotspot/share/utilities/nativeCallStack.cpp

Those changes seem to be transient artefacts of the workflow and should disappear on their own, eventually.

The reason those changes appeared in this PR is that this PR's branch is based on a more recent master than that of the PR that this PR depends on, https://github.com/openjdk/jdk/pull/16461.

**Comment by jddarcy** [comment]
@pavelrappo , please create a CSR for this behavioral change.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Utils.java [hunk 1]
```diff
@@ -310,6 +311,17 @@ public boolean isProperty(String name) {
...
+     * If a similar query is added to javax.lang.model, use that instead.
+     */
+    public static boolean isImplicitlyDeclaredClass(Tree e) {
+        if (e.getKind() != Tree.Kind.CLASS)
+            return false;
+        if (!(e instanceof com.sun.tools.javac.tree.JCTree.JCClassDecl classDecl))
+            return false; // this should probably never happen with jdk.compiler
+        return (classDecl.mods.flags & com.sun.tools.javac.code.Flags.IMPLICIT_CLASS) != 0;
+    }
+
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Utils.java:324]: Generally, this should all be in the `Workarounds` class, which is the intended enclosure for any code that needs to access `javac` internals that is not directly available via any public API.

Alternatively, you could put this now into `Trees` (not `java.lang.model`, since this is an AST-only method, not an `Element`-based method.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclint/Checker.java [hunk 2]
```diff
@@ -102,6 +102,7 @@
 import jdk.javadoc.internal.doclint.HtmlTag.AttrKind;
 import jdk.javadoc.internal.doclint.HtmlTag.ElemKind;
 import static jdk.javadoc.internal.doclint.Messages.Group.*;
+import jdk.javadoc.internal.doclets.toolkit.util.Utils;
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclint/Checker.java:105]: This (ugly) dependency would go away if you put that new method into the `Trees` API.

File: test/langtools/jdk/javadoc/doclet/testImplicitlyDeclaredClasses/TestImplicitlyDeclaredClasses.java [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+        new TestImplicitlyDeclaredClasses().runTests();
+    }
+
+    @Test
+    public void test(Path base) throws IOException {
+        int i = 0;
+        for (Method main : mainMethods())
+            for (Method otherMethod : otherMethods()) {
+                var methods = List.of(main, otherMethod);
+                String index = String.valueOf(i++);
```
Comment (jonathan-gibbons) [inline test/langtools/jdk/javadoc/doclet/testImplicitlyDeclaredClasses/TestImplicitlyDeclaredClasses.java:61]: Why `String`, not `var` ?

File: test/langtools/jdk/javadoc/doclet/testImplicitlyDeclaredClasses/TestImplicitlyDeclaredClasses.java [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+                //
+                // Here's how this check works: if an undocumented class warning
+                // is present, then the total count of undocumented element warnings
+                // is one greater than that of undocumented methods.
+                //
+                // Of course, it's possible, although seemingly unlikely, that
+                // this check passes, when it should fail: the warning for class
+                // is generated, but the warning for the first method is not.
+                // Numbers are equal, test passes.
+                checking("");
```
Comment (jonathan-gibbons) [inline test/langtools/jdk/javadoc/doclet/testImplicitlyDeclaredClasses/TestImplicitlyDeclaredClasses.java:117]: Better to give a non-empty string identifying the nature of the check

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Utils.java [hunk 5]
```diff
@@ -310,6 +311,17 @@ public boolean isProperty(String name) {
         return options.javafx() && name.endsWith("Property");
     }
 
+    /*
+     * If a similar query is added to javax.lang.model, use that instead.
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Utils.java:315]: The `Trees` API would be a better place than `java.lang.model` since this is just a query on an AST node, not on an `Element`.

**Comment by pavelrappo** [comment]
@jonathan-gibbons I pushed a few updates.

While I agree with your stance on Utils here, I don't think it's for JavaDoc to define anything language-related. Have a look and tell me if what you see in those updates addresses your concerns.

**Review by jonathan-gibbons — APPROVED** [review]
TL;DR: approved.

"One of these days" we should try and split the `jdk.javadoc` module in two: one for the "tool" part, and another for the "standard doclet" part. 

In that split world, the "tool" part is deeply intertwined with the `javac` front end, and gets to access `javac` front-end internals almost without restriction. As such, it does not need to use `WorkArounds`.   In module terms, it would be given qualified access into `jdk.compiler` internals. And, in that split world", the doclet part would/could/should be a pure client of the Language Model API, to reflect over the declarations, and the Compiler Tree API, to access the documentation comments. And, for that module, `WorkArounds` is used to workaround shortcomings of those public API.  So it would reluctantly be given "temporary" access to `javac` internals, until the workarounds can be eliminated.

All of which is to say that `ElementsTable` will have an unnecessary forward dependency on `WorkArounds`, which belongs in that hypothetical "other" module. But at least the method  the method has a well placed comment. `DocLint` is using access to `javac` internals, but `DocLint` is currently something of an orphan stepchild of `javac` and `javadoc` and deserves more significant remedial attention of its own, so gets a pass for now.

**Review by jonathan-gibbons — APPROVED** [review]
Latest version approved

**Review by jonathan-gibbons — APPROVED** [review]
